File: solutions/python/isbn-verifier/1/isbn_verifier.py
def is_valid(isbn):
    digits = [chr for chr in isbn if chr != "-"]
    if len(digits) != 10:
        return False
    if digits[-1] == "X":
       digits[-1] = "10"
    sum, mult = 0, 10
    print(digits)
    for chr in digits:
        if not chr.isdecimal():
            return False
        sum += mult * int(chr)
        mult -= 1
    return sum % 11 == 0

File: solutions/haskell/isbn-verifier/1/src/IsbnVerifier.hs
module IsbnVerifier (isbn) where

import Data.Char (isDigit, digitToInt)

isbn :: String -> Bool
isbn str =
    let
        noDashes = filter (/='-') str
        numbers = map (charToNum) noDashes
    in
        if length numbers /= 10
          || last numbers > 10
          || any (>9) (take 9 numbers)
          then False
          else check numbers
    where
        check numbers = rem (sum (zipWith (*) numbers [10,9..1])) 11 == 0

charToNum :: Char -> Int
charToNum 'X' = 10
charToNum chr = if isDigit chr then digitToInt chr else 99